### ISSUE TITLE
Unable to connect to WinRM using domain account from computer not joined into the domain

### ISSUE THREAD

**Issue by ssbarnea** [ansible/ansible#13416]
I am struggling to make WinRM work with ansible 1.9.4 but it seems impossible. Ansible runs from Linux machines and we need to execute some actions on Windows machines as a domain account. 

The linux hosts are not joined into the domain as that's clearly not an option.

```
+ ansible citrite -vvvv --module-name ping -i inventory.ini
<winadmin.do.example.com> ESTABLISH WINRM CONNECTION FOR USER: svcacct_xxx on PORT 5985 TO winadmin.do.example.com
<winadmin.do.example.com> WINRM CONNECT: transport=plaintext endpoint=http://winadmin.do.example.com:5985/wsman
<winadmin.do.example.com> WINRM CONNECTION ERROR: 401 Unauthorized.
<winadmin.do.example.com> WINRM CONNECT: transport=plaintext endpoint=https://winadmin.do.example.com:5985/wsman
<winadmin.do.example.com> WINRM CONNECTION ERROR: 500 WinRMTransport. [SSL: UNKNOWN_PROTOCOL] unknown protocol (_ssl.c:590)
winadmin.do.example.com | FAILED => 500 WinRMTransport. [SSL: UNKNOWN_PROTOCOL] unknown protocol (_ssl.c:590)
```

The variables are:

```
ansible_ssh_user: svcacct_xxx
ansible_ssh_pass: xxx
ansible_ssh_port: 5985
ansible_connection: winrm
ansible_winrm_scheme: http
```

I mention that WinRM is configured to allow HTTP connections.

```
C:\WINDOWS\system32>winrm get winrm/config
Config
    Client
        NetworkDelayms = 5000
        URLPrefix = wsman
        AllowUnencrypted = true
        Auth
            Basic = true
            Digest = true
            Kerberos = true
            Negotiate = true
            Certificate = true
            CredSSP = false
        DefaultPorts
            HTTP = 5985
```

I also tried to use the domain notation for user, as in `svcacct_xxx@EXAMPLE.COM` but in this case it still fails because it does try to use kerberos, which cannot work.

In fact it seems that it does something that does not make much sense to me: it gets a 401 response from the server over HTTP port but after this it tries to connect using SSL on the same port?!

**Comment by nitzmahone** [comment]
Currently, kerberos is the only way to do domain auth (due to restrictions in the Python WinRM client we use). Kerberos auth over WinRM requires that the Linux host:
- be able to access your AD DNS infrastructure (as default resolver, or at least domain resolver for your AD domain)
- have a valid computer account in the AD domain
- have these details configured properly in the Linux kerberos setup

This stuff is very limited right now- we're working on adding support for NTLM and CredSSP to the Python WinRM client, which should allow domain auth over WinRM and eliminate the need for the Linux host to be configured for Kerberos. 

There isn't a definitive one-stop guide yet, but take a look at these two resources to get going:

https://community.spiceworks.com/how_to/44885-setup-centos-to-authenticate-via-active-directory (you can skip the system-auth part at the end if you don't want Linux SSO from Active Directory)

http://hindenes.com/trondsworking/2015/07/27/ansible-and-windows-configuring-kerberosactive-directory-authentication/

**Comment by jhawkesworth** [comment]
Hopefully the instructions here are of some assistance too: http://docs.ansible.com/ansible/intro_windows.html#active-directory-support

You don't have to have your ansible controller joined to the kerberos domain (in the sense that you log into the controller using domain username and password), you can just use kinit to acquire a kerberos ticket before running your playbooks.

Hope this helps,

Jon

**Comment by mss** [comment]
As @jhawkesworth noted does the `winrm` module work if you get a kerberos token via `kinit` before executing ansible, even if the host isn't joined to the domain. It fails though if you try to rely on the `ansible_user`/`ansible_password` combination.

I'm not sure if this is an issue with ansible not passing the credentials properly or if this is an issue in pywinrm itself and is filed as diyan/pywinrm#36 already. Whatever it is, it should be fixable though since getting a kerberos ticket obviously works.

The [documentation](http://docs.ansible.com/ansible/intro_windows.html#active-directory-support) should also be made clearer insofar as that a `kinit` is not only for testing purposes but a required command right now. (I just noticed that the docs are on GitHub as well…)

```
[root@host ansible]# ansible --version
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

[root@host ansible]# kdestroy -A
[root@host ansible]# ansible -i inventory windows -m win_ping -vvvvv
Using /etc/ansible/ansible.cfg as config file
Loaded callback minimal of type stdout, v2.0
<win.example.com> ESTABLISH WINRM CONNECTION FOR USER: ucadmin@EXAMPLE.COM on PORT 5986 TO win.example.com
<win.example.com> WINRM CONNECT: transport=kerberos endpoint=https://win.example.com:5986/wsman
<win.example.com> WINRM CONNECTION ERROR: authGSSClientInit() failed: (('Unspecified GSS failure.  Minor code may provide more information', 851968), ("Can't find client principal ucadmin@EXAMPLE.COM in cache collection", -1765328243))
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/winrm.py", line 152, in _winrm_connect
    self.shell_id = protocol.open_shell(codepage=65001) # UTF-8
  File "/usr/lib/python2.7/site-packages/winrm/protocol.py", line 132, in open_shell
    res = self.send_message(xmltodict.unparse(req))
  File "/usr/lib/python2.7/site-packages/winrm/protocol.py", line 207, in send_message
    return self.transport.send_message(message)
  File "/usr/lib/python2.7/site-packages/winrm/transport.py", line 170, in send_message
    prepared_request = self.session.prepare_request(request)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 394, in prepare_request
    hooks=merge_hooks(request.hooks, self.hooks),
  File "/usr/lib/python2.7/site-packages/requests/models.py", line 299, in prepare
    self.prepare_auth(auth, url)
  File "/usr/lib/python2.7/site-packages/requests/models.py", line 494, in prepare_auth
    r = auth(self)
  File "/usr/lib/python2.7/site-packages/requests_kerberos/kerberos_.py", line 318, in __call__
    auth_header = self.generate_request_header(None, host, is_preemptive=True)
  File "/usr/lib/python2.7/site-packages/requests_kerberos/kerberos_.py", line 158, in generate_request_header
    raise KerberosExchangeError("%s failed: %s" % (kerb_stage, str(error.args)))
KerberosExchangeError: authGSSClientInit() failed: (('Unspecified GSS failure.  Minor code may provide more information', 851968), ("Can't find client principal ucadmin@EXAMPLE.COM in cache collection", -1765328243))

<win.example.com> WINRM CONNECT: transport=ssl endpoint=https://win.example.com:5986/wsman
<win.example.com> WINRM CONNECTION ERROR: the specified credentials were rejected by the server
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/winrm.py", line 152, in _winrm_connect
    self.shell_id = protocol.open_shell(codepage=65001) # UTF-8
  File "/usr/lib/python2.7/site-packages/winrm/protocol.py", line 132, in open_shell
    res = self.send_message(xmltodict.unparse(req))
  File "/usr/lib/python2.7/site-packages/winrm/protocol.py", line 207, in send_message
    return self.transport.send_message(message)
  File "/usr/lib/python2.7/site-packages/winrm/transport.py", line 179, in send_message
    raise InvalidCredentialsError("the specified credentials were rejected by the server")
InvalidCredentialsError: the specified credentials were rejected by the server

win.example.com | UNREACHABLE! => {
    "changed": false, 
    "msg": "kerberos: authGSSClientInit() failed: (('Unspecified GSS failure.  Minor code may provide more information', 851968), (\"Can't find client principal ucadmin@EXAMPLE.COM in cache collection\", -1765328243)), ssl: the specified credentials were rejected by the server", 
    "unreachable": true
}
```

```
[root@host ansible]# kinit ucadmin@EXAMPLE.COM
Password for ucadmin@EXAMPLE.COM: 
[root@zmsync ansible]# ansible -i inventory windows -m win_ping -vvvvv
Using /etc/ansible/ansible.cfg as config file
Loaded callback minimal of type stdout, v2.0
<win.example.com> ESTABLISH WINRM CONNECTION FOR USER: ucadmin@EXAMPLE.COM on PORT 5986 TO win.example.com
<win.example.com> WINRM CONNECT: transport=kerberos endpoint=https://win.example.com:5986/wsman
<win.example.com> WINRM OPEN SHELL: E27CE2C3-EA03-4E9B-A7C8-A3240D759F60
<win.example.com> EXEC Set-StrictMode -Version Latest
(New-Item -Type Directory -Path $env:temp -Name "ansible-tmp-1466514099.29-144497244938395").FullName | Write-Host -Separator '';
<win.example.com> WINRM EXEC u'PowerShell' [u'-NoProfile', u'-NonInteractive', u'-ExecutionPolicy', u'Unrestricted', u'-EncodedCommand', u'UwBlAHQALQBTAHQAcgBpAGMAdABNAG8AZABlACAALQBWAGUAcgBzAGkAbwBuACAATABhAHQAZQBzAHQACgAoAE4AZQB3AC0ASQB0AGUAbQAgAC0AVAB5AHAAZQAgAEQAaQByAGUAYwB0AG8AcgB5ACAALQBQAGEAdABoACAAJABlAG4AdgA6AHQAZQBtAHAAIAAtAE4AYQBtAGUAIAAiAGEAbgBzAGkAYgBsAGUALQB0AG0AcAAtADEANAA2ADYANQAxADQAMAA5ADkALgAyADkALQAxADQANAA0ADkANwAyADQANAA5ADMAOAAzADkANQAiACkALgBGAHUAbABsAE4AYQBtAGUAIAB8ACAAVwByAGkAdABlAC0ASABvAHMAdAAgAC0AUwBlAHAAYQByAGEAdABvAHIAIAAnACcAOwA=']
<win.example.com> WINRM RESULT u'<Response code 0, out "C:\\Users\\ucadmin\\App", err "#< CLIXML\r\n<Objs Ver">'
<win.example.com> PUT "/tmp/tmp_EBV0Y" TO "C:\Users\ucadmin\AppData\Local\Temp\ansible-tmp-1466514099.29-144497244938395\win_ping.ps1"
<win.example.com> WINRM PUT "/tmp/tmp_EBV0Y" to "C:\Users\ucadmin\AppData\Local\Temp\ansible-tmp-1466514099.29-144497244938395\win_ping.ps1" (offset=8847 size=8847)
<win.example.com> EXEC Set-StrictMode -Version Latest
Try
{
& 'C:\Users\ucadmin\AppData\Local\Temp\ansible-tmp-1466514099.29-144497244938395\win_ping.ps1'
}
Catch
{
$_obj = @{ failed = $true }
If ($_.Exception.GetType)
{
$_obj.Add('msg', $_.Exception.Message)
}
Else
{
$_obj.Add('msg', $_.ToString())
}
If ($_.InvocationInfo.PositionMessage)
{
$_obj.Add('exception', $_.InvocationInfo.PositionMessage)
}
ElseIf ($_.ScriptStackTrace)
{
$_obj.Add('exception', $_.ScriptStackTrace)
}
Try
{
$_obj.Add('error_record', ($_ | ConvertTo-Json | ConvertFrom-Json))
}
Catch
{
}
Echo $_obj | ConvertTo-Json -Compress -Depth 99
Exit 1
}
Finally { Remove-Item "C:\Users\ucadmin\AppData\Local\Temp\ansible-tmp-1466514099.29-144497244938395" -Force -Recurse -ErrorAction SilentlyContinue }
<win.example.com> WINRM EXEC u'PowerShell' [u'-NoProfile', u'-NonInteractive', u'-ExecutionPolicy', u'Unrestricted', u'-EncodedCommand', u'UwBlAHQALQBTAHQAcgBpAGMAdABNAG8AZABlACAALQBWAGUAcgBzAGkAbwBuACAATABhAHQAZQBzAHQACgBUAHIAeQAKAHsACgAmACAAJwBDADoAXABVAHMAZQByAHMAXAB1AGMAYQBkAG0AaQBuAFwAQQBwAHAARABhAHQAYQBcAEwAbwBjAGEAbABcAFQAZQBtAHAAXABhAG4AcwBpAGIAbABlAC0AdABtAHAALQAxADQANgA2ADUAMQA0ADAAOQA5AC4AMgA5AC0AMQA0ADQANAA5ADcAMgA0ADQAOQAzADgAMwA5ADUAXAB3AGkAbgBfAHAAaQBuAGcALgBwAHMAMQAnAAoAfQAKAEMAYQB0AGMAaAAKAHsACgAkAF8AbwBiAGoAIAA9ACAAQAB7ACAAZgBhAGkAbABlAGQAIAA9ACAAJAB0AHIAdQBlACAAfQAKAEkAZgAgACgAJABfAC4ARQB4AGMAZQBwAHQAaQBvAG4ALgBHAGUAdABUAHkAcABlACkACgB7AAoAJABfAG8AYgBqAC4AQQBkAGQAKAAnAG0AcwBnACcALAAgACQAXwAuAEUAeABjAGUAcAB0AGkAbwBuAC4ATQBlAHMAcwBhAGcAZQApAAoAfQAKAEUAbABzAGUACgB7AAoAJABfAG8AYgBqAC4AQQBkAGQAKAAnAG0AcwBnACcALAAgACQAXwAuAFQAbwBTAHQAcgBpAG4AZwAoACkAKQAKAH0ACgBJAGYAIAAoACQAXwAuAEkAbgB2AG8AYwBhAHQAaQBvAG4ASQBuAGYAbwAuAFAAbwBzAGkAdABpAG8AbgBNAGUAcwBzAGEAZwBlACkACgB7AAoAJABfAG8AYgBqAC4AQQBkAGQAKAAnAGUAeABjAGUAcAB0AGkAbwBuACcALAAgACQAXwAuAEkAbgB2AG8AYwBhAHQAaQBvAG4ASQBuAGYAbwAuAFAAbwBzAGkAdABpAG8AbgBNAGUAcwBzAGEAZwBlACkACgB9AAoARQBsAHMAZQBJAGYAIAAoACQAXwAuAFMAYwByAGkAcAB0AFMAdABhAGMAawBUAHIAYQBjAGUAKQAKAHsACgAkAF8AbwBiAGoALgBBAGQAZAAoACcAZQB4AGMAZQBwAHQAaQBvAG4AJwAsACAAJABfAC4AUwBjAHIAaQBwAHQAUwB0AGEAYwBrAFQAcgBhAGMAZQApAAoAfQAKAFQAcgB5AAoAewAKACQAXwBvAGIAagAuAEEAZABkACgAJwBlAHIAcgBvAHIAXwByAGUAYwBvAHIAZAAnACwAIAAoACQAXwAgAHwAIABDAG8AbgB2AGUAcgB0AFQAbwAtAEoAcwBvAG4AIAB8ACAAQwBvAG4AdgBlAHIAdABGAHIAbwBtAC0ASgBzAG8AbgApACkACgB9AAoAQwBhAHQAYwBoAAoAewAKAH0ACgBFAGMAaABvACAAJABfAG8AYgBqACAAfAAgAEMAbwBuAHYAZQByAHQAVABvAC0ASgBzAG8AbgAgAC0AQwBvAG0AcAByAGUAcwBzACAALQBEAGUAcAB0AGgAIAA5ADkACgBFAHgAaQB0ACAAMQAKAH0ACgBGAGkAbgBhAGwAbAB5ACAAewAgAFIAZQBtAG8AdgBlAC0ASQB0AGUAbQAgACIAQwA6AFwAVQBzAGUAcgBzAFwAdQBjAGEAZABtAGkAbgBcAEEAcABwAEQAYQB0AGEAXABMAG8AYwBhAGwAXABUAGUAbQBwAFwAYQBuAHMAaQBiAGwAZQAtAHQAbQBwAC0AMQA0ADYANgA1ADEANAAwADkAOQAuADIAOQAtADEANAA0ADQAOQA3ADIANAA0ADkAMwA4ADMAOQA1ACIAIAAtAEYAbwByAGMAZQAgAC0AUgBlAGMAdQByAHMAZQAgAC0ARQByAHIAbwByAEEAYwB0AGkAbwBuACAAUwBpAGwAZQBuAHQAbAB5AEMAbwBuAHQAaQBuAHUAZQAgAH0A']
<win.example.com> WINRM RESULT u'<Response code 0, out "{"changed":false,"pi", err "#< CLIXML\r\n<Objs Ver">'
<win.example.com> WINRM CLOSE SHELL: E27CE2C3-EA03-4E9B-A7C8-A3240D759F60
win.example.com | SUCCESS => {
    "changed": false, 
    "invocation": {
        "module_name": "win_ping"
    }, 
    "ping": "pong"
}
```

**Comment by jhawkesworth** [comment]
Its not really a bug in ansible that you need a valid kerberos ticket to use kerberos to communicate with windows hosts.  People have different needs and different infrastructures and if you set up your ansible controller to use active directory to authorize your users, then once you have logged in you will have a kerberos ticket.  If you don't want your linux ansible controller to use active directory for logins then you have to acquire a ticket in another way.  You can run kinit manually or use a wrapper script to get it before you kick off ansible.  Or you can use a callback plugin to acquire the ticket before you start (which is what I do - code for it is here https://github.com/ansible/ansible/pull/14972 although it needs some revision to use against latest ansible versions).

SSH Keys and kerberos tickets are rather different things - I put a bit about this in my ansiblefest talk which is available here: https://www.ansible.com/ansible-2-windows

This feature idea is rather old now and things have moved on.  If you install pywinrm 0.2.0 you can use NTLM v2 for auth too.

**Comment by mss** [comment]
I am fine with requiring a ticket via `kinit` even though a working `ansible_password` variable with a password in vault would be nice. I looked into the pykerberors code and it seems that there's no way to acquire a new ticket via that codebase so `kinit` (or domain membership) is probably the only sensible way to go.

I think there are two bugs here though:
- The documentation is misleading and suggests that you don't need a ticket via `kinit` (unless for testing) and you could `ansible_password` which cost me quite some time until I figured out that the issue was on ansible's side.
- Right now the `ansible_user` is not really used as a username but as a trigger to enable kerberos and if I remember the code correctly does pywinrm's API always require some credentials so it is a bit of a workaround for that as well. In the end always the principal of you previously acquired kerberos ticket is used. While the latter makes sense, the overall behaviour is confusing.

**Comment by jhawkesworth** [comment]
I know @nitzmahone has been working on the windows documentation recently so I'll mention so he can take your feedback about the documentation into consideration.

**Comment by dariusjs** [comment]
@mss  Since the update to 2.1 on an OS X system of and on a CentOS7 machine the ansible_user does not seem to trigger a kerberos look up anymore. I have another system on the older 2.0.2 code which works fine still. 

With the 2.0.0.2 version I would do a kinit with my own username and could winrm to the windows hosts. With 2.1 I can only winrm when doing a kinit with the username specified in the config. 

`node3.local | UNREACHABLE! => {
    "changed": false, 
    "msg": "kerberos: authGSSClientInit() failed: (('Unspecified GSS failure.  Minor code may provide more information', 851968), (\"Can't find client principal ansible@NET.LOCAL in cache collection\", -1765328243)), ssl: auth method ssl requires a password", 
    "unreachable": true
}`


The documentation http://docs.ansible.com/ansible/intro_windows.html here is what we followed for our 2.0.0.2 setup 

> If you have installed the kerberos module and ansible_user contains @ (e.g. username@realm), Ansible will first attempt Kerberos authentication. This method uses the principal you are authenticated to Kerberos with on the control machine and not ``ansible_user``. If that fails, either because you are not signed into Kerberos on the control machine or because the corresponding domain account on the remote host is not available, then Ansible will fall back to “plain” username/password authentication.

**Comment by nitzmahone** [comment]
The new docs will be much clearer on these topics (as well as how/why to select a particular auth type). 

The reason for ansible_user being ignored was actually a bug in pykerberos that they recently accepted my fix for, so if you have the newest Ansible/pywinrm/pykerberos/requests-kerberos, you can use as many kerberos creds as you want in parallel with ansible_user from inventory, so long as you have valid TGTs cached for all of them.

**Comment by nitzmahone** [comment]
As the underlying issue is a non-issue, and the other things are addressed in forthcoming docs, going to close this one.
